rfd/0080-yubikey-piv.md

```diff
@@ -0,0 +1,272 @@
+
+If the slot is already in use, we will provide the user with a prompt to reuse this key, overwrite it, or abort. Users can also provide the `--yubikey-overwrite` or `--yubikey-reuse` flags to bypass this prompt.
+
+```
+> tsh login
+
+yubikey slot 9a has a pre-generated private key with public key:
+-----BEGIN PUBLIC KEY-----
+...
+-----END PUBLIC KEY-----
+
+Would you like to:
+ (1) Resuse this key
+ (2) Overwrite this key
+ (3) Abort login
+```
+
+In the case of a Teleport user connecting to multiple clusters, they'll need to use `--yubikey-reuse` in order to maintain access to them simultaneously.
+
+Note that reusing the key is not any less secure than generating a new one, as the Teleport Auth server can still attest that the key was generated on the yubikey, and therefore cannot be exported. 
+
+#### new `tsh` flags
+
+| Name | Default Value(s) | Allowed Value(s) | Description |
+| `-y, --yubikey` | - | - | During login, use yubikey to generate and store private key data for your login session. |
+| `--yubikey-pin-policy` | never | always once never | Can be provided with `tsh login -y` to control how often you need to enter your yubikey PIN to access the yubikey private key data. |
+| `--yubikey-touch-policy` | never | always cached never | Can be provided with `tsh login -y` to control how often you need to touch your yubikey to access the yubikey private key data. |
+| `--yubikey-overwrite` | - | - | When provided, `tsh login -y` will overwrite existing yubikey data if it exists. |
+| `--yubikey-reuse` | - | - | When provided, `tsh login -y` will reuse existing yubikey data if it exists. Useful when using yubikey to connect to multiple clusters simultaneously. |
+
+#### tsh config file
+
+Users can also provide their yubikey login settings in their tsh config file (`/etc/tsh.yaml` or `~/.tsh/config/config.yaml`). This is important so that users don't need to provide the flag options on every login.
+
+```yaml
+yubikey: true
+yubikey_pin: once
+yubikey_touch: cached
+yubikey_reuse: true
+```
+
+#### Multiple yubikeys
+
+If the user has multiple yubikeys connected, they will receive an error asking them to connect just one yubikey. Otherwise, we can't guarentee that the right yubikey will be chosen. We may be able to add support for this in the future.
+
+### Additional considerations
+
+- At first, yubikey login support will only be added to `tsh`, but in the future other clients like `tctl` and the WebUI can support it with the same general flow. 
+  - When `yubikey_login_required` is enforced on the Auth Server, we can expect other Teleport clients to fail to login unless we add yubikey login support for them as well. Initial implementation will most likely suffer from this deficiency.
```

Comment:
My 2c: I'd rather tweak our flag meanings so users can't shoot themselves in the foot. We can push for stricter settings when enough of the systems are ready for it.

Reply:
This is a feature requirement, so it will need to be included even in preview mode when not all clients support it. I agree it's not ideal, but we can do our best to communicate this to users, especially since it will be an opt-in feature in preview mode.